Re-introduce the 1.4 counters API

russelldb opened the pull request:
Re-add the 1.4 API.

Support legacy counters at the FSM, Vnode, kv_crdt level.

I really think you don't need to review `riak_kv_gcounter` and `riak_kv_pncounter` as they are practically unchanged from 1.4 (The 2.0 API requires `update` returns {ok, Val} is the only change.)

The upgrade support is untested as upgrade seems to be broken. But running legacy clients against 2.0 has been tested (just run the riak_test for counters.)

russelldb committed:
Re-add 1.4 counter code. Make counters in 2.0 backwards compatible

russelldb committed:
Update v1 counter for v2 API

russelldb committed:
Add a no-op inc/dec by zero to pncounter

Update test to use legacy mod
@TODO add tests for othe mods/generalize kv_crdt_eqc

russelldb committed:
V1 binary encode kv counters

Use capabilities in pb service

lenary commented on include/riak_kv_types.hrl:
```diff
@@ -5,6 +5,8 @@
 
 %% Top Level Key->Type Types
+-define(LEGACY_COUNTER_TYPE, riak_kv_pncounter).
+-define(LEGACY_COUNTER_TYPE(Val), #crdt{mod=?LEGACY_COUNTER_TYPE, ctype="application/riak_counter", value=Val}).
```
I'd prefer "V1_COUNTER_TYPE" just in case another type comes along.... but I'm relatively easy about it.

lenary commented on test/kv_crdt_eqc.erl:
```diff
@@ -324,7 +324,7 @@ pncounterds() ->
 pncounter() ->
     ?LET(PNCounter, pncounterds(),
-         riak_kv_crdt:to_binary(?CRDT{mod=riak_dt_pncounter, value=PNCounter}, ?V1_VERS)).
+         riak_kv_crdt:to_binary(?CRDT{mod=riak_kv_pncounter, value=PNCounter}, ?V1_VERS)).
```
I think this should be `?LEGACY_COUNTER_TYPE`

seancribbs commented on src/riak_kv_web.erl:
```diff
@@ -96,9 +96,9 @@ raw_dispatch(Name) ->
-     {["buckets", bucket, crdt, key], {riak_kv_wm_crdt, known_type},
-      riak_kv_wm_crdt, Props2}
+     %% counters
+     {["buckets", bucket, counters, key],
```
`counters` should be a string unless you intend to bind the value of that path segment.

engelsanchez replied:
I suppose there will be some overlap with the URL configuration for this and upcoming HTTP API for CRDTs that @seancribbs and @russelldb have discussed? So for now going back to "counters" (instead of counters atom) should be enough?

engelsanchez commented on src/riak_kv_gcounter.erl:
```diff
@@ -0,0 +1,258 @@
+%% that both counters contain the same actors and those actors have the same count.
+-spec equal(gcounter(), gcounter()) -> boolean().
+equal(VA,VB) ->
+    lists:sort(VA) =:= lists:sort(VB).
```
It would be nice to add this method in the eunit test

engelsanchez replied:
Ignore my previous comment. I was looking at riak_test coverage. Ignore. Ignore.

engelsanchez commented:
I noticed that the riak_test only exercises counters through the HTTP interface. Is there a test that uses the PB interface? If not, could we tweak the riak test to use it sometimes?

russelldb commented:
The riak_test verify_counter_capability uses the PB interface as does verify_dt_capability, but the latter fails due to upgrading to 2.0 failing.

russelldb commented:
verify_dt_converge uses the new PB interface. I guess you want a counter test that runs against 2.0 that uses the legacy 1.4 message? We'd have that if upgrade worked.

russelldb committed:
Remove bucket type support from legacy API

You can have that in the new API

engelsanchez commented on src/riak_kv_crdt.erl:
```diff
@@ -98,16 +98,22 @@ crdt_value(Type, {ok, {_Meta, ?CRDT{mod=Type, value=Value}}}) ->
     lager:info("Error(s) deserializing CRDT at ~p ~p: ~p~n", [Bucket, Key, Errors]).
 
+maybe_log_sibling_crdts(Bucket, Key, CRDTs) when length(CRDTs) > 1 ->
+    lager:info("Sibling CRDTs at ~p ~p: ~p~n", [Bucket, Key, orddict:fetch_keys(CRDTs)]);
```
Should this be an error or warning? I think so. Also the above would work better as one too. I missed that before.

russelldb replied:
The second one is probably an error, as it means some kind of failure in our application of bucket types to CRDTs. The first one is tough to decide…it will happen if CRDT is serialized incorrectly or corrupt, but it will also happen if a users data as a binary _happens_ to start with the series of bytes that a CRDT does. It would be a big  coincidence I guess, but since EQC found a few…I dunno…pick one, you're the reviewer :smile:

engelsanchez replied:
I get to pick? oh, oh. Well I think it's better to log it as an error then. If someone is using non-CRDT data as CRDT and they hit the serialization magic numbers, that is errorish enough right there.

engelsanchez commented on src/riak_kv_crdt.erl:
```diff
@@ -148,12 +154,23 @@ deserialize_crdt(?V2_VERS, CRDTBin) ->
     {error, {invalid_version, V}}.
 
+counter_op(N) when N < 0 ->
+    {decrement, N*-1};
```
Kind of a long winded way of writing -N yo

russelldb replied:
:blush:

engelsanchez commented on src/riak_kv_pb_counter.erl:
```diff
@@ -39,8 +39,9 @@
--include_lib("riak_pb/include/riak_kv_pb.hrl").
--include_lib("riak_pb/include/riak_pb_kv_codec.hrl").
+-include_lib("../../riak_pb/include/riak_kv_pb.hrl").
+-include_lib("../../riak_pb/include/riak_pb_kv_codec.hrl").
```
Forgot to remove the ../../ bit?

russelldb replied:
Again, bollocks.

engelsanchez commented on src/riak_kv_pb_counter.erl:
```diff
@@ -71,11 +69,11 @@ decode(Code, Bin) ->
-        #rpbcountergetreq{type=T, bucket=B} ->
-            Bucket = bucket_type(T, B),
+        #rpbcountergetreq{bucket=B} ->
+            Bucket = bucket_type(B),
```
Is there a reason to translate to {<<"default">>, Bucket} here? I thought all the code downstream from here could handle the old style, but I could be wrong.

russelldb replied:
@Vagabond  added it, I don't know the answer.

Vagabond replied:
Either should be valid... I think... The 2-tuple form might be safer.

engelsanchez commented:
I don't have anything else to add for the moment. Ping me when you address the existing comments.

russelldb committed:
Add support for using 1.4 style counters via the 2.0 API

Still not sure of the argument as to why, but it was compelling
enough that I did the work.

russelldb committed:
Address review comments

russelldb commented:
When you "existing comments" you want the riak_tests writing too? If not, ping @engelsanchez

russelldb commented:
Sorry for the hefty refactor on riak_kv_pb_crdt, but adding another error (bucket type not found) just made the case nesting on the two `process` funs another level deep.

engelsanchez commented:
@russelldb would you mind merging from develop? There are non-trivial merge conflicts in riak_kv_web.erl and riak_kv_pb_crdt.erl (around the section you refactored heavily).

engelsanchez commented:
for the merge: in riak_kv_web.erl the counters descriptor has to be reworked now that Sean has added descriptors for URLs including bucket types (which counters don't need). And in riak_kv_pb_crdt.erl, in develop they were added the retry_put_coordinator_failure request option set to false.

russelldb committed:
Don't return riak_kv_pb_counter state for a crdt message

russelldb committed:
Merge branch 'develop' into feature/rdb/count-back

Conflicts:
	src/riak_kv_pb_crdt.erl
	src/riak_kv_web.erl

russelldb committed:
Fix merge error

russelldb commented:
OK, did the merges.

I know you said maybe remove the roundtrip 1.4->2.0->1.4 stuff for counters, but I think this poosh answers the questions. If not then I will just cut it out for now. Going to work on a roundtrip test in the meantime.

russelldb commented:
Also, related https://github.com/basho/riak_test/pull/418 (verify_crdt_capability)

engelsanchez commented on src/riak_kv_pb_crdt.erl:
```diff
@@ -164,6 +108,112 @@ process_stream(_,_,State) ->
+    make_option(w, W) ++
+        make_option(dw, DW) ++
+        make_option(pw, PW) ++
+        make_option(timeout, timeout(Timeout)) ++
```
@russelldb the merge missed adding the retry_put_coordinator_failure option present in develop, added by 0449970

russelldb replied:
it is on line 168, it is an exceptional option, and not something passed in (we mandate false for crdts, since counters are not idempotent we want users to choose about retrying.) However, good point if you mention that the 1.4 adapter stuff doesn't do it…I mean if 2.0 counters opt out of fsm retry, 1.4 counters should too (I consider this something Scott missed but I should add it to this PR)

engelsanchez commented on src/riak_kv_web.erl:
```diff
@@ -104,6 +105,10 @@ raw_dispatch(Name) ->
+     %% counters
+     {["buckets", bucket, "counters", key],
+      riak_kv_wm_counter, Props},
+
```
Notice that this now adds the entry for counters twice, as this section is inside a list comprehension for two entries: one with types prefix and one without. I don't know if that has any side effects. It might be OK. But ideally this should be outside of it, adding only one entry.

russelldb replied:
it is? Where is the other entry?

engelsanchez replied:
Notice that this section is now within the list comprehension that ends here: https://github.com/basho/riak_kv/blob/ebf28d94cf7de8d78e3214cad5a5dc528efc22e3/src/riak_kv_web.erl#L115

russelldb replied:
Oh I see. Ugly.

russelldb replied:
I hope adding it twice is fine, I don't know though though, more and more I'm realising functional programming and no side effects aren't really the same thing. What can I say, what test we have pass.

russelldb replied:
So riak_kv_web is basically wrong with regards to counters. There is no v3 or v1 api for counters, only v2. I guess it only passed review because I removed counters for ages to make 2.0 dev easier. More fool me.

engelsanchez replied:
These are just properties passed to the module in the webmachine route. So, if the counters module doesn't care about these api or other properties, they are just ignored. Meh.

russelldb replied:
Looking at https://github.com/basho/riak_kv/blob/ebf28d94cf7de8d78e3214cad5a5dc528efc22e3/src/riak_kv_web.erl#L113 is it wrong? I thought CRDTs cannot be stored in the empty bucket type, that there is no default path for them @seancribbs?

engelsanchez commented on src/riak_kv_pb_crdt.erl:
```diff
@@ -215,3 +272,79 @@ get_context(Ctx, true) ->
+process_legacy_counter({error, Reason}, State) ->
+    {error, {format, Reason}, State};
+process_legacy_counter(Req, State) ->
+    LegacyState = riak_kv_pb_counter:init(),
```
This works fine here. It does create a riak_client on every request, which maybe we can optimize later.

engelsanchez commented:
Ok, I finally got a setup with all relevant tests passing (with 1.3.2 as legacy and 1.4.2 as previous). Sorry for the delay but I managed to completely botch my riak_test setup while preparing for that.  I think at this point the only remaining issue is the FSM retry flag for counters which you wanted to address here.

russelldb committed:
Add new fsm retry option to counters

Do not allow bucket type operations on v1.4 counters
TODO: there seems to be a <<"default">> bucket prefix path added
for CRDTs, I think this is a mistake.

russelldb commented:
Addressed comments cc @engelsanchez. I've pinged @seancribbs about that prefix + prefixless path for CRDTs.

seancribbs commented:
The way you have it is fine. We should remove the other resource's dispatch line until it's ready, as such it has not been tested adequately and we haven't thoroughly designed the API.

russelldb committed:
Remove CRDT HTTP API endpoint, there isn't one yet

russelldb commented:
Removed!

engelsanchez commented:
Latest changes look good to me, relevant riak tests still pass and coverage is good.

:+1: :boat: :dancer:

Vagabond commented:
I think this PR forgot to add security checks to the WM interface.

russelldb commented:
Bum. Will start a new branch / PR for that. Thanks @Vagabond
